src/com/projectkorra/projectkorra/util/DamageHandler.java

```diff
@@ -28,64 +29,76 @@
 	 * @param damage The amount of damage to deal
 	 */
 	public static void damageEntity(final Entity entity, Player source, double damage, final Ability ability, boolean ignoreArmor) {
-		if (entity instanceof LivingEntity && TempArmor.hasTempArmor((LivingEntity) entity)) {
-			ignoreArmor = true;
-		}
 		if (ability == null) {
 			return;
 		}
+		
+		if (entity instanceof LivingEntity) {
+			if (((LivingEntity) entity).getNoDamageTicks() > 0) {
+				return;
+			}
+			
+			if (TempArmor.hasTempArmor((LivingEntity) entity)) {
+				ignoreArmor = true;
+			}
+		}
+		
 		if (source == null) {
 			source = ability.getPlayer();
 		}
 
 		final AbilityDamageEntityEvent damageEvent = new AbilityDamageEntityEvent(entity, ability, damage, ignoreArmor);
+		
+		if (entity instanceof Player && Commands.invincible.contains(entity.getName())) {
+			damageEvent.setCancelled(true);
+		}
+		
 		Bukkit.getServer().getPluginManager().callEvent(damageEvent);
-		if (entity instanceof LivingEntity) {
-			if (entity instanceof Player && Commands.invincible.contains(entity.getName())) {
-				damageEvent.setCancelled(true);
+		
+		if (entity instanceof LivingEntity && !damageEvent.isCancelled()) {
+			LivingEntity lent = (LivingEntity) entity;
+			damage = Math.max(0, damageEvent.getDamage());
+			
+			if (damageEvent.doesIgnoreArmor() && damage > 0) {
+				double defense = lent.getAttribute(Attribute.GENERIC_ARMOR).getValue();
+				double toughness = lent.getAttribute(Attribute.GENERIC_ARMOR_TOUGHNESS).getValue();
+				damage /= 1 - (Math.min(20, Math.max(defense / 5, defense - 4 * damage / (toughness + 8)))) / 25;
```

Comment:
What's happening here? What are these numbers?

Reply:
This is all from the formula for how Minecraft calculates damage given the presence of armor, just happening in reverse so we can calculate what our damage input needs to be so that the final damage stays the same as what was passed to `DamageHandler#damageEntity(...)`

Reply:
Ok, maybe a comment would be good?

Reply:
I don't really think a comment is needed here, it's just some basic algebra and the context is more than enough to figure out what's happening

Reply:
I couldn't figure it out and it wouldn't hurt to add a comment explaining where those numbers came from, but I'm not a pk dev so you can do whatever you want.